Kernel/FileSystem/FATFS/Inode.cpp

```diff
@@ -55,26 +55,116 @@ ErrorOr<Vector<BlockBasedFileSystem::BlockIndex>> FATInode::compute_block_list()
+        // the "FAT ID" field with some flags, Cluster 1 should be the end of chain
+        // marker).
+        // Internally, we use `cluster == 0` to represent the root directory Inode,
+        // which is a signal to read the root directory region blocks on FAT12/16
+        // file systems. (`fs().first_block_of_cluster` will return the appropriate
+        // block/sectors to read given cluster == 0).
+        // Therefore, we read one set of sectors for these invalud cluster numbers,
+        // and then terminate the loop becuase the FAT entry at `cluster` for these
+        // values does not represent the next step in the chain (because there is
+        // nothing else to read).
+        if (cluster <= 1) {
+            break;
+        }
 
-        u32 fat_offset = cluster * sizeof(u32);
-        u32 fat_sector_index = fs().boot_record()->reserved_sector_count + (fat_offset / fs().m_device_block_size);
+        u32 fat_offset = fat_offset_for_cluster(cluster);
+        u32 fat_sector_index = fs().m_parameter_block->block->reserved_sector_count + (fat_offset / fs().m_device_block_size);
         u32 entry_offset = fat_offset % fs().m_device_block_size;
 
         TRY(fs().raw_read(fat_sector_index, fat_sector_buffer));
 
-        cluster = *reinterpret_cast<u32*>(&fat_sector->data()[entry_offset]);
-        cluster &= cluster_number_mask;
+        // Look up the next cluster to read, or read End of Chain marker from table.
+        cluster = cluster_number(*fat_sector, cluster, entry_offset);
     }
 
     return block_list;
 }
 
+u32 FATInode::no_more_clusters() const
+{
+    // Returns the end of chain entry for the given file system.
+    // Any FAT entry of this value or greater signifies the end
+    // of the chain has been reached for a given entry.
+    switch (fs().m_fat_version) {
+    case FATVersion::FAT12:
+        return 0xFF8;
+    case FATVersion::FAT16:
+        return 0xFFF8;
+    case FATVersion::FAT32:
+        return 0x0FFFFFF8;
+    default:
+        VERIFY(false);
+    }
+}
+
+size_t FATInode::fat_offset_for_cluster(u32 cluster) const
+{
+    switch (fs().m_fat_version) {
+    case FATVersion::FAT12: {
+        // In FAT12, a cluster entry is stored in a byte, plus
+        // the low/high nybble of an adjacent byte.
+        //
+        // CLSTR:   0 1      2 3      4 5
+        // INDEX: [0 1 2], [3 4 5], [6 7 8]
+
+        // Every 2 clusters are represented using 3 bytes.
+        return (cluster * 3) / 2;
+    } break;
+    case FATVersion::FAT16:
+        return cluster * 2; // Each cluster is stored in 2 bytes.
+    case FATVersion::FAT32:
+        return cluster * 4; // Each cluster is stored in 4 bytes.
+    default:
+        VERIFY(false);
+    }
+}
+
+u32 FATInode::cluster_number(KBuffer const& fat_sector, u32 entry_cluster_number, u32 entry_offset) const
+{
+    u32 cluster = 0;
+    switch (fs().m_fat_version) {
+    case FATVersion::FAT12: {
+        u8 fat16_byte1;
+        u8 fat16_byte2;
```

Comment:
initialize variables at first use, or always initialize to 0.

```
u8 fat16_byte1 = fat_sector.bytes(entry_offset + 0)
```

If the bytes are stored strangely, then possibly the LittleEndian and BigEndian types in AK can be helpful.

Reply:
Thank you, this definitely cleaned things up!